### PR TITLE
Reuse Manager amongst http calls

### PR DESCRIPTION
**Pull request by charlescrain**
# Changes

* Provider typeclass removed, server uri passed into `runWeb3'` 
* Web3 now contains ReaderT

### CODE REVIEW

File: src/Network/Ethereum/Web3/JsonRpc.hs [hunk 1]
```diff
@@ -24,36 +24,35 @@ import           Network.Ethereum.Web3.Types
...
+                                                 RequestBody (RequestBodyLBS),
+                                                 httpLbs, method, parseRequest,
+                                                 requestBody, requestHeaders,
+                                                 responseBody)
 
 -- | Name of called method.
 type MethodName = Text
 
--- | JSON-RPC server URI
-type ServerUri  = String
```
Comment (akru) [inline src/Network/Ethereum/Web3/JsonRpc.hs]: You drop `ServerUri` here, but use in line 67.

File: src/Network/Ethereum/Web3/Provider.hs [hunk 2]
```diff
@@ -14,30 +14,24 @@ module Network.Ethereum.Web3.Provider where
 import           Control.Concurrent.Async    (Async, async)
 import           Control.Exception           (try)
 import           Control.Monad.IO.Class      (MonadIO (..))
+import           Control.Monad.Trans.Reader  (mapReaderT, runReaderT)
 import           Network.Ethereum.Web3.Types
-
--- | Ethereum node service provider
-class Provider a where
```
Comment (akru) [inline src/Network/Ethereum/Web3/Provider.hs]: I'm not sure that static configuration is better than `Provider` typeclass, ok.

File: src/Network/Ethereum/Web3/Provider.hs [hunk 3]
```diff
@@ -14,30 +14,24 @@ module Network.Ethereum.Web3.Provider where
...
-data DefaultProvider
-
-instance Provider DefaultProvider where
-    rpcUri = return "http://localhost:8545"
+import           Network.HTTP.Client         (newManager)
+import           Network.HTTP.Client.TLS     (tlsManagerSettings)
 
 -- | 'Web3' monad runner
-runWeb3' :: MonadIO m => Web3 a b -> m (Either Web3Error b)
+runWeb3' :: MonadIO m => ServerUri -> Web3 a -> m (Either Web3Error a)
```
Comment (akru) [inline src/Network/Ethereum/Web3/Provider.hs]: Just define `Web3Config` datatype with `Default` instance.

File: src/Network/Ethereum/Web3/Provider.hs [hunk 4]
```diff
@@ -14,30 +14,24 @@ module Network.Ethereum.Web3.Provider where
...
+runWeb3' uri f = do
+    manager <- liftIO $ newManager tlsManagerSettings
+    liftIO . try .  flip runReaderT (uri, manager) . unWeb3 $ f
 
 -- | 'Web3' runner for default provider
-runWeb3 :: MonadIO m => Web3 DefaultProvider b -> m (Either Web3Error b)
+runWeb3 :: MonadIO m => Web3 a -> m (Either Web3Error a)
 {-# INLINE runWeb3 #-}
-runWeb3 = runWeb3'
+runWeb3 = runWeb3' "http://localhost:8545"
```
Comment (akru) [inline src/Network/Ethereum/Web3/Provider.hs]:
```haskell
runWeb3 = runWeb3' def
```

File: src/Network/Ethereum/Web3/Types.hs [hunk 5]
```diff
@@ -31,10 +32,15 @@ import           GHC.Generics
 import           Network.Ethereum.Unit
 import           Network.Ethereum.Web3.Address           (Address, zero)
 import           Network.Ethereum.Web3.Encoding.Internal (toQuantityHexText)
-import           Network.Ethereum.Web3.Internal          (toLowerFirst)
+import           Network.Ethereum.Web3.Internal
+import           Network.HTTP.Client                     (Manager)
+
+
+-- | JSON-RPC server URI
+type ServerUri  = String
```
Comment (akru) [inline src/Network/Ethereum/Web3/Types.hs]: Oh, it's here, okay.

File: src/Network/Ethereum/Web3/Provider.hs (hunk 3 above)
Comment (charlescrain) [reply src/Network/Ethereum/Web3/Provider.hs]: for the Web3Config, are you looking for something like 
```haskell
newtype Web3Config = Web3Config { unWeb3Config :: ServerUri }
```

Comment (akru) [reply src/Network/Ethereum/Web3/Provider.hs]: Yes, to improve performance we can use `newtype` here.

But, in my opinion, more generic way is define something like

```haskell
data Provider = HttpProvider ServerUri | IpcProvider FilePath
  deriving (Show, Eq, Generic)

data Web3Config = Web3Config {
  provider :: Provider
} deriving (Show, Eq, Generic)

Comment (charlescrain) [reply src/Network/Ethereum/Web3/Provider.hs]: Okay, seems straightforward.  Seems like we could scrap `Web3Config` and just use `Provider` defining the Default instance on `Provider`

Comment (charlescrain) [reply src/Network/Ethereum/Web3/Provider.hs]: Also, my current implementation doesn't support Ipc. The example you posted, https://kseo.github.io/posts/2017-01-23-custom-connection-manager-for-http-client.html  requires a URI and a `FilePath` to the socket. I see `FilePath` is used to create the `Manager` but I'm not 100% sure what the `ServerUri` would be. For now I would like to define the Provider type as
```haskell
-- TODO: Add to type `IpcProvider FilePath` to support IPC
data Provider = HttpProvider ServerUri 
  deriving (Show, Eq, Generic)
```
 and add  support for IPC to the roadmap. How does this sound?

Comment (akru) [reply src/Network/Ethereum/Web3/Provider.hs]: Cool!

Also, may it'll be interested to add ability for users to pass custom `Manager`, for example:

```haskell
data Provider = HttpProvider ServerUri | CustomProvider Manager
  deriving (Show, Eq, Generic)
```

**Comment by charlescrain** [comment]
Made changes to include Provider type

**Comment by akru** [comment]
Perfect! Thank you for your work!

**Pull request merged**